### ISSUE TITLE
missing scaler.scale in manual_backward?

### ISSUE THREAD

**Issue by AljoSt** [PyTorchLightning/pytorch-lightning#4148]
## 🐛 Bug

When using `manual_backward` and `precision=16`, an exception is raised:
RuntimeError: unscale_() has already been called on this optimizer since the last update().

As far as I can see, the scaler is actually never used to scale the loss during `manual_backward` and in the example [here](https://pytorch-lightning.readthedocs.io/en/latest/optimizers.html) it is not mentioned that one has to do it before.
Also, I cannot find any 'update' call

colab model to reproduce:

https://colab.research.google.com/drive/1n7-NxC0IJ3gYJGZLsoeOtu-oQc3I8UjF?usp=sharing

**Comment by alpapado** [comment]
I am also facing this issue when code identical to the example with 16 bit precision

**Comment by edenlightning** [comment]
might be fixed by https://github.com/PyTorchLightning/pytorch-lightning/pull/4311

**Comment by edenlightning** [comment]
@tchaton please update status if you made any progress on this

**Comment by SeanNaren** [comment]
I can give an update here, it seems that within manual backward with native AMP we are missing some calls to the scaler, particular the `update` call. This happens because we never call `optimizer_step` in the accelerator when using manual_backward.

To remedy this, you can add this to the bottom of your override training step:

```python
# scale when native amp
if self.trainer.amp_backend == AMPType.NATIVE:
       self.trainer.scaler.update()
```

such as:

```python
        def training_step(self, batch, batch_idx, optimizer_idx):
            # manual
            (opt_a, opt_b) = self.optimizers()
            loss_1 = self.step(batch[0])

            # make sure there are no grads
            if batch_idx > 0:
                assert torch.all(self.layer.weight.grad == 0)

            self.manual_backward(loss_1, opt_a)
            opt_a.step()
            opt_a.zero_grad()
            assert torch.all(self.layer.weight.grad == 0)

            # fake discriminator
            loss_2 = self.step(batch[0])

            # ensure we forward the correct params to the optimizer
            # without retain_graph we can't do multiple backward passes
            self.manual_backward(loss_2, opt_b, retain_graph=True)
            self.manual_backward(loss_2, opt_a, retain_graph=True)

            assert self.layer.weight.grad is not None
            opt_b.step()
            opt_b.zero_grad()
            assert torch.all(self.layer.weight.grad == 0)
            
            # scale when native amp
            if self.trainer.amp_backend == AMPType.NATIVE:
                 self.trainer.scaler.update()
```

The work/discussion is going on in https://github.com/PyTorchLightning/pytorch-lightning/pull/4485

**Comment by edenlightning** [comment]
@tchaton

**Comment by SeanNaren** [comment]
This has been solved in #4485 with the addition of manual_optimizer_step:

```python
def training_step(...):
                (opt_a, opt_b) = self.optimizers()
                loss = ...
                # automatically applies scaling, etc...
                self.manual_backward(loss, opt_a)
                # This will force an opt.step() even if accumulate_grad_batches is set.
                self.manual_optimizer_step(opt_a, force_optimizer_step=True, custom_args=4)
```

manual_optimizer step will ensure appropriate step across hardware, and does the scaler update as well!
